patroni/quorum.py

```diff
@@ -0,0 +1,316 @@
+import logging
+
+from typing import Collection, Iterator, Optional, Tuple
+
+from .collections import CaseInsensitiveSet
+
+logger = logging.getLogger(__name__)
+
+
+class QuorumError(Exception):
+    pass
+
+
+class QuorumStateResolver(object):
+    """Calculates a list of state transition tuples of the form `('sync'/'quorum'/'restart',leader,number,set_of_names)`
+
+    Synchronous replication state is set in two places. PostgreSQL configuration sets how many and which nodes are
+    needed for a commit to succeed, abbreviated as `numsync` and `sync` set here. DCS contains information about how
+    many and which nodes need to be interrogated to be sure to see an xlog position containing latest confirmed commit,
+    abbreviated as `quorum` and `voters` set. Both pairs have the meaning "ANY n OF set".
+
+    The number of nodes needed for commit to succeed, `numsync`, is also called the replication factor.
+
+    To guarantee zero lost transactions on failover we need to keep the invariant that at all times any subset of
+    nodes that can acknowledge a commit overlaps with any subset of nodes that can achieve quorum to promote a new
+    leader. Given a desired replication factor and a set of nodes able to participate in sync replication there
+    is one optimal state satisfying this condition. Given the node set `active`, the optimal state is:
+
+        sync = voters = active
+        numsync = min(sync_wanted, len(active))
+        quorum = len(active) - numsync
+
+    We need to be able to produce a series of state changes that take the system to this desired state from any
+    other state arbitrary given arbitrary changes is node availability, configuration and interrupted transitions.
+
+    To keep the invariant the rule to follow is that when increasing `numsync` or `quorum`, we need to perform the
+    increasing operation first. When decreasing either, the decreasing operation needs to be performed later.
+
+    Order of adding or removing nodes from sync and voters depends on the state of synchronous_standby_names:
```

Comment:
Why does it depend on `synchronous_standby_names`?
(Sorry if I'm putting too many questions, I'm still going over files 😅 Maybe some of them will be answered when I take a look somewhere else).